prober/derp.go

```diff
@@ -577,6 +580,196 @@ func runDerpProbeNodePair(ctx context.Context, from, to *tailcfg.DERPNode, fromc
+var derpProbeBandwidthMu sync.Mutex
+
+// derpProbeBandwidthTUN takes two DERP clients (fromc and toc) connected to two
+// DERP servers (from and to) and sends a test payload of a given size from one
+// to another over TUN.
+func derpProbeBandwidthTUN(ctx context.Context, transferTime *expvar.Float, from, to *tailcfg.DERPNode, fromc, toc *derphttp.Client, size int64) error {
+	derpProbeBandwidthMu.Lock()
+	defer derpProbeBandwidthMu.Unlock()
+
+	// Temporarily set up a TUN device with which to simulate a real client TCP connection
+	// tunneling over DERP.
+	dev, _, err := tstun.New(log.Printf, "utun9")
+	if err != nil {
+		return fmt.Errorf("failed to create TUN device: %w", err)
+	}
+	defer dev.Close()
+	mtu, err := dev.MTU()
+	if err != nil {
+		return fmt.Errorf("failed to get TUN MTU: %w", err)
+	}
+
+	r, err := router.New(log.Printf, dev, nil, nil)
+	if err != nil {
+		return fmt.Errorf("failed to create router: %w", err)
+	}
+	defer r.Close()
+
+	err = r.Set(&router.Config{
+		LocalAddrs: []netip.Prefix{netip.MustParsePrefix("172.16.0.1/30")},
+		Routes:     []netip.Prefix{netip.MustParsePrefix("172.16.0.1/30")},
+	})
+	if err != nil {
+		return fmt.Errorf("failed to set routes: %w", err)
+	}
+
+	// Read packets from TUN device
+	tunReadC := make(chan error, 1)
+	go func() {
+		// There's not a lot of thought in the below number. 5 doesn't use a ton of memory,
+		// and a bigger number like 20 didn't seem to increase throughput.
+		numBufs := 5
+		bufs := make([][]byte, 0, numBufs)
+		sizes := make([]int, numBufs)
+		for range numBufs {
+			bufs = append(bufs, make([]byte, mtu+tstun.PacketStartOffset))
+		}
+
+		for {
+			n, err := dev.Read(bufs, sizes, tstun.PacketStartOffset)
+			if err != nil {
+				return
+			}
+
+			for i := range n {
+				buf := bufs[i][:sizes[i]+tstun.PacketStartOffset]
+				pkt := buf[tstun.PacketStartOffset:]
+				src, dst := make([]byte, 4), make([]byte, 4)
+				srcPort, dstPort := make([]byte, 2), make([]byte, 2)
+				copy(src, pkt[12:16])
+				copy(dst, pkt[16:20])
+				// TODO: the below assumes no IP options in header
+				copy(srcPort, pkt[20:22])
+				copy(dstPort, pkt[22:24])
+
+				// Swap
+				copy(pkt[12:], dst)
+				copy(pkt[16:], src)
+				copy(pkt[20:], dstPort)
+				copy(pkt[22:], srcPort)
+				if err := fromc.Send(toc.SelfPublicKey(), buf); err != nil {
+					tunReadC <- err
+					return
+				}
+			}
+		}
+	}()
+
+	// Receive the packets.
+	recvc := make(chan error, 1)
+	go func() {
+		defer close(recvc) // to break out of 'select' below.
+		for {
+			m, err := toc.Recv()
+			if err != nil {
+				recvc <- fmt.Errorf("failed to receive: %w", err)
+				return
+			}
+			switch v := m.(type) {
+			case derp.ReceivedPacket:
+				if v.Source != fromc.SelfPublicKey() {
+					recvc <- fmt.Errorf("got data packet from unexpected source, %v", v.Source)
+					return
+				}
+				pkt := v.Data
+				dev.Write([][]byte{pkt}, tstun.PacketStartOffset)
```

Comment:
Write assumes the offset is free scratch space, is it?

Reply:
Sorry, I'm not sure what you mean by this? What's the significance?

Reply:
take a look at https://github.com/tailscale/wireguard-go/blob/4e883d38c8d363e92e02445e4c664f2c27ffdb88/tun/tun_linux.go#L376

we need to put this scratch space in front of `pkt` rather than stealing from it

Reply:
pkt has free scratch space at the beginning, it's actually not just the TCP packet, it's the buffer that we read from the TUN device earlier, including the scratch space. Does that address your concern?